File: lexer.cpp
#include<iostream>
#include<string>
#include<map>
#include<ctype.h>
#include<cstdio>
namespace lexer{
	enum Tag{
		NONE=0,AND=256,BASIC,BREAK,DO,ELSE,EQ,FALSE,GE,ID,IF,INDEX,LE,MINUS,NE,NUM,OR,REAL,TEMP,TRUE,WHILE
	};
	class Token;
	class Lexer;
	class Num;
	class Word;
	class Real;
}
namespace symbols{
	
}

using namespace symbols;
using namespace lexer;

namespace lexer{
	class Token{
		public:
			int tag;
			Token():tag(Tag::NONE){
			}
			Token(int t):tag(t){
			}
			std::string toString(){
				std::string s;
				return s+(char)tag;
			}
	};
	class Num:public Token{
		public:
			int val;
			Num():Token(),val(0){
			}
			Num(int v):Token(Tag::NUM),val(v){
			}
			std::string toString()
			{
				std::string s;
				return s+std::to_string(val);
			}
	};
	class Word:public Token{
		public:
			std::string lexeme;
			Word():Token(),lexeme("none"){
			}
			Word(std::string s,int tag):Token(tag),lexeme(s){
			}
			std::string toString()
			{
				return lexeme;
			}
			static const Word And,Or,Eq,Ne,Le,Ge,Minus,True,False,Temp;
	};
	class Real:public Token{
		public:
			double val;
			Real():Token(),val(0.0){
			}
			Real(double v):Token(Tag::REAL),val(v){
			}
			std::string toString()
			{
				std::string s;
				return s+std::to_string(val);
			}
	};
	class Lexer{
		public:
			static int line;
			char peek;
			std::map<std::string,Word> words;
			void reserve(Word w){
				words.insert(std::pair<std::string,Word>(w.lexeme,w));
			}
			Lexer():peek(' '){
				reserve(Word("if",Tag::IF));
				reserve(Word("else",Tag::ELSE));
				reserve(Word("while",Tag::WHILE));
				reserve(Word("do",Tag::DO));
				reserve(Word("break",Tag::BREAK));
				reserve(Word::True);reserve(Word::False);
		//		reserve(Type::Int);reserve(Type::Char);
		//		reserve(Type::Bool);reserve(Type::Float);
			}
			void readch(){
				peek=getchar();
			}
			bool readch(char c){
				readch();
				if(peek!=c)	return false;
				peek=' ';
				return true;
			}
			Token* scan(){
				for(;;readch())
				{
					if(peek==' '||peek=='\t')continue;
					else if(peek=='\n')line++;
					else break;
				}
				switch(peek){
					case '&':{
						if(readch('&'))return new Word(Word::And);
						else return new Token('&');
						break;
					}
					case '|':{
						if(readch('|'))return new Word(Word::Or);
						else return new Token('|');
						break;
					}
					case '=':{
						if(readch('='))return new Word(Word::Eq);
						else return new Token('=');
						break;
					}
					case '!':{
						if(readch('='))return new Word(Word::Ne);
						else return new Token('!');
						break;
					}
					case '<':{
						if(readch('='))return new Word(Word::Le);
						else return new Token('<');
						break;
					}
					case '>':{
						if(readch('='))return new Word(Word::Ge);
						else return new Token('>');
						break;
					}
				}
				if(isdigit(peek)){
					int v=0;
					do{
						v=10*v+peek-'0';
						readch();
					}while(isdigit(peek));
					if(peek!='.') return new Num(v);
					double x=v,d=10.0;
					while(true)
					{
						readch();
						if(!isdigit(peek))
							break;
						x+=(peek-'0')/d;
						d*=10.0;
					}
					return new Real(x);
				}
				if(isalpha(peek))
				{
					std::string b;
					do{
						b+=peek;
						readch();
					}while(isalnum(peek));
					
					Word w=words[b];
					if(w.lexeme!="none")return new Word(w);
					Word *id=new Word(b,Tag::ID);
					words.insert(std::pair<std::string,Word>(b,*id));
					
					return id;
				}
				
				Token* tok=new Token(peek);
				peek=' ';
				return tok;
			}		
	};

	int Lexer::line=1;
	const Word Word::And("&&",Tag::AND),Word::Or("||",Tag::OR),
	           Word::Eq("==",Tag::EQ),Word::Ne("!=",Tag::NE),
			   Word::Le("<=",Tag::LE),Word::Ge(">=",Tag::GE),
			   Word::Minus("minus",Tag::MINUS),
			   Word::True("true",Tag::TRUE),
			   Word::False("false",Tag::FALSE),
			   Word::Temp("t",Tag::TEMP);
}
namespace symbols{
	class Env{
		public:
			
	};
}

int main()
{

	Lexer L;
	while(true)
	{
		Token* tmp=L.scan();
		std::cout<<tmp->tag<<std::endl;
		delete tmp;
	}
}

File: Lexer.py
# -*- coding: utf-8 -*-
"""
Created on Sun Jun 21 01:17:27 2020

@author: HP
"""

Key_word=["if","else","while","do","break","true","false"]
Type=["int","char","bool","float"]
Logical=["true","false"]
Split=["\t"," ","\n"]

def Lexer(src):
    obj=[]
    for line in src:
        out=[]
        i=0
        while i<len(line):
            i,token=scan(i,line)
            if i!=len(line)+1:
                out.append(token)
        obj.append(out)
    return obj

def scan(i,line):
    while i<len(line):
        if line[i] in Split:
            i+=1
            continue
        else:
            break
    if i==len(line):
        return i+1,("",)
    if line[i]=='&':
        if readch(i+1,line,'&'):
            return i+2,("&&",)
        else:
            return i+1,("&",)
    elif line[i]=='|':
        if readch(i+1,line,"|"):
            return i+2,("||",)
        else:
            return i+1,('|',)
    elif line[i]=='=':
        if readch(i+1,line,"="):
            return i+2,("==",)
        else:
            return i+1,('=',)
    elif line[i]=='>':
        if readch(i+1,line,"="):
            return i+2,(">=",)
        else:
            return i+1,('>',)
    elif line[i]=='<':
        if readch(i+1,line,"="):
            return i+2,("<=",)
        else:
            return i+1,('<',)
    elif line[i]=='!':
        if readch(i+1,line,"="):
            return i+2,("!=",)
        else:
            return i+1,('!',)
    elif getch(i,line) and getch(i,line).isdigit():
        v=0
        while getch(i,line) and getch(i,line).isdigit():
            v=10*v+int(getch(i,line))
            i+=1
        if getch(i,line) and readch(i,line,'.'):
            d=0.1
            i+=1
            while getch(i,line) and getch(i,line).isdigit():
                v+=d*int(getch(i,line))
                i+=1
                d*=0.1
            return i,("real",v)
        else:
            return i,("num",v)
    elif getch(i,line) and (getch(i,line).isalpha() or getch(i,line)=='_') :
        tmp=""
        tmp+=getch(i,line)
        i+=1
        while getch(i,line) and (getch(i,line).isalnum() or getch(i,line)=='_'):
            tmp+=getch(i,line)
            i+=1
        if tmp in Key_word+Logical:
            return i,(tmp,)
        elif tmp in Type:
            return i,("basic",tmp)
        else:
            return i,("id",tmp)
    else:
        return i+1,(line[i],)        
        
def readch(i,line,ch):
    if i<len(line) and line[i]==ch:
        return True
    else:
        return False

def getch(i,line):
    if i<len(line):
        return line[i]
    else:
        return False
if __name__=="__main__":
    with open("test.txt","r") as f:
        src=f.readlines()
        obj=Lexer(src)
        for items in obj:
            for item in items:
                print(item,end="\t")
            print()